microprofile/lra/coordinator/src/main/java/org/wildfly/extension/microprofile/lra/coordinator/service/LRARecoveryService.java

```diff
@@ -36,6 +36,8 @@ public synchronized void start(final StartContext context) throws StartException
                     .noneMatch(rm -> rm instanceof LRARecoveryModule)) {
                     RecoveryManager.manager().addModule(lraRecoveryModule);
                 }
+                // run an initial recovery scan - skip periodicWorkFirstPass since it is a no-op
+                lraRecoveryModule.periodicWorkSecondPass();
```

Comment:
@mmusgrov I expect this blocks and thus unless @ropalka says otherwise it should use the MSC asynchronous API (StartContext.asynchronous() and from a runnable executed by an executor, StartContext.complete().

See the class and start method javadoc at https://github.com/jboss-msc/jboss-msc/blob/main/src/main/java/org/jboss/msc/Service.java

There are lots of examples of this in the WF code. The general pattern is to inject an executor that the server kernel provides into your service instance:

https://github.com/wildfly/wildfly/blob/main/batch-jberet/src/main/java/org/wildfly/extension/batch/jberet/deployment/BatchEnvironmentProcessor.java#L144-L146

and then use it in start

https://github.com/wildfly/wildfly/blob/main/batch-jberet/src/main/java/org/wildfly/extension/batch/jberet/job/repository/JdbcJobRepositoryService.java#L45-L61

Reply:
It can potentially block for up to [two seconds](https://github.com/jbosstm/lra/blob/0.0.10.Final/service-base/src/main/java/io/narayana/lra/LRAConstants.java#L66) for each participant if recovery needs to reissue JAX RS requests to participants. I'll update the PR along the lines you suggest ...

Reply:
I did all that but the StartContext.complete() call results in an IllegalStateException: Illegal controller state:

```
    private void startRecoveryScan(final StartContext context) {
        assert executorSupplier != null;
        final ExecutorService service = executorSupplier.get();
        final Runnable task = () -> {
            try {
                // run an initial recovery scan - skip periodicWorkFirstPass since it is a no-op
                lraRecoveryModule.periodicWorkSecondPass();
                context.complete(); // results in java.lang.IllegalStateException: Illegal controller state
            } catch (Exception e) {
                MicroProfileLRACoordinatorLogger.LOGGER.failedToRunRecoveryScan(CONTEXT_PATH, e);
            }
        };
        try {
            service.execute(task);
        } catch (RejectedExecutionException e) {
            task.run();
        } finally {
            context.asynchronous();
        }
    }
```
If I comment out context.complete(); it works - am I calling it in the correct place, I modelled the call based on your [last link](https://github.com/wildfly/wildfly/blob/main/batch-jberet/src/main/java/org/wildfly/extension/batch/jberet/job/repository/JdbcJobRepositoryService.java#L45-L61
).

Reply:
Ah wait, I was calling `context.complete();` elsewhere too - let me retest that ...